### PR TITLE
improve int data type check in session

### PR DESCRIPTION
**Pull request by yupbank**
right now this would fail
```
import tensorflow as tf
sess = tf.InteractiveSession()
tf.placeholder(tf.string, name='t').eval(feed_dict={'t_1:0': 1})
```
with  
```
TypeError: object() takes no parameters
```

### CODE REVIEW

File: tensorflow/python/framework/dtypes.py [hunk 1]
```diff
@@ -127,6 +127,11 @@ def as_numpy_dtype(self):
     """Returns a `numpy.dtype` based on this `DType`."""
     return _TF_TO_NP[self._type_enum]
 
+  def convert_to_numpy_obj(self, obj):
+    """Explicitly convert obj based on numpy type except for string type."""
+    numpy_dtype = self.as_numpy_dtype
+    return numpy_dtype(obj) if  numpy_dtype is not object else str(obj)
```
Comment (mrry) [inline tensorflow/python/framework/dtypes.py]: Can you move this to be a private function in `session.py`? I'd prefer not to add it to the public API.

File: tensorflow/python/client/session.py [hunk 2]
```diff
@@ -1073,9 +1073,9 @@ def _feed_fn(feed, feed_val):
 
           subfeed_dtype = subfeed_t.dtype.as_numpy_dtype
           if isinstance(subfeed_val,
-                        int) and subfeed_dtype(subfeed_val) != subfeed_val:
+                        int) and subfeed_t.dtype.convert_to_numpy_obj(subfeed_val) != subfeed_val:
```
Comment (mrry) [inline tensorflow/python/client/session.py]: Can you add a test case to `session_test.py` that exercises this new code?

**Review by mrry — CHANGES_REQUESTED** [review]

File: tensorflow/python/framework/dtypes.py (hunk 1 above)
Comment (yupbank) [reply tensorflow/python/framework/dtypes.py]: but then.. we need to be careful not to add more `np.object` like types which... might not happen, but if.. then the session won't notice .. 
but anyway, i have addressed the comments

File: tensorflow/python/client/session_test.py [hunk 3]
```diff
@@ -1742,6 +1742,12 @@ def testAddFunctionToGrpcSession(self):
     server = server_lib.Server.create_local_server()
     self.runTestAddFunctionToSession(server.target)
 
+  def testAutoConvertAndCheckData(self):
+    with self.test_session() as sess:
+      a = array_ops.placeholder(tf.string)
+      with self.assertRaises(TypeError) as e:
```
Comment (mrry) [inline tensorflow/python/client/session_test.py]: Style nit: can you please change this line to the following and remove line 1750?

```python
with self.assertRaisesRegexp(TypeError, 'Type of feed value 1 with type in is not'):
```

**Review by mrry — APPROVED** [review]
Looks good... thanks for the contribution!

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by martinwicke** [comment]
@yupbank session_test is failing, can you look at this?

**Comment by yupbank** [comment]
ah... i made a mistake by specifying the sting type use tf directly..

**Comment by yupbank** [comment]
@tensorflow-jenkins test this please.

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by yupbank** [comment]
@tensorflow-jenkins test this please.

**Comment by martinwicke** [comment]
Jenkins, test this please.

slice_op_test is broken.

**Comment by yupbank** [comment]
weird error messages...

i don't know what's broken or how can i even help ?

**Comment by martinwicke** [comment]
Looks like the tests here are all good. @sb2nov this is ready to merge.

**Comment by caisq** [comment]
@tensorflow-jenkins test this please

**Pull request merged**